Add support for Omega's `PseudoThickness` variable

xylar opened the pull request:
<!--
Thank you for your pull request.
Please add a description of what is accomplished in the PR here at the top:
-->

# Summary

This PR updates Polaris to track a set of variable renames in Omega that clarify the distinction between pseudo-thickness (the prognostic z-star/z-tilde coordinate variable) and true geometric layer thickness. The renames affect both the MPAS-Ocean → Omega variable mapping and the forward-run output specifications. A secondary fix corrects two bugs in the merry-go-round test exposed by this work.

## Implementation details

### Variable renames (`mpaso_to_omega.yaml` and forward YAMLs)

The following Omega variable names changed, and the MPAS-Ocean → Omega mapping plus any task-specific forward YAML output lists are updated accordingly:

| Old name | New name |
|---|---|
| `LayerThickness` | `PseudoThickness` |
| `VerticalVelocity` | `VerticalPseudoVelocity` |
| `TotalVerticalVelocity` | `TotalVerticalPseudoVelocity` |
| `BottomDepth` | `BottomGeomDepth` |
| `ZMid` | `GeomZMid` |
| `ZInterface` | `GeomZInterface` |

The pseudo/geometric thickness conversion in `ocean_model_step.py` was reworked so that `PseudoThickness` and `SpecVol` are written directly, and `GeomLayerThickness` is computed on read-back rather than assumed to be stored.

### Z-tilde rework (`polaris/ocean/vertical/ztilde.py`)

The z-tilde vertical coordinate was rewritten to be independent of z-star infrastructure. Previously the two shared code paths, but after the renaming this became unnecessarily complicated. `ztilde.py` now has its own self-contained implementation, and `horiz_press_grad` (the only task using z-tilde) was updated accordingly.

### Merry-go-round fixes

Two bugs in the merry-go-round test were fixed:
1. **Initial state conversion** — incorrect variable was used when converting the initial state.
2. **Constant EOS** — the task now registers a constant equation of state, which is required for the test to run correctly under Omega.

### Omega submodule

Update Omega submodule to [54d85d8c4f](https://github.com/E3SM-Project/Omega/tree/54d85d8c4fd24e969af8bb4764d108df9caaa0c5)
This bring in:
* https://github.com/E3SM-Project/Omega/pull/335
* https://github.com/E3SM-Project/Omega/pull/381
* https://github.com/E3SM-Project/Omega/pull/400
* https://github.com/E3SM-Project/Omega/pull/414
* https://github.com/E3SM-Project/Omega/pull/327

<!--
Below are a few things we ask you or your reviewers to kindly check. 
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist
* [x] User's Guide has been updated
* [x] Developer's Guide has been updated
* [x] API documentation in the Developer's Guide (`api.md`) has any new or modified class, method and/or functions listed
* [x] Documentation has been [built locally](https://e3sm-project.github.io/polaris/main/developers_guide/building_docs.html) and changes look as expected
* [x] `Testing` comment in the PR documents testing used to verify the changes

<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

xylar committed:
Update mpas-ocean --> omega map

Switch layerThickness --> GeometricLayerThickness

Add a few vertical coordinate variables

xylar committed:
Update Omega CTest meshes

The new meshes have as many variables as possible mapped to
the new Omega names, and include PseudoThickness (if they
previously had layerThickness) as well as GeometricLayerThickness.
The PseudoThickness is computed using TEOS-10 for the specific
volume.

xylar commented on utils/omega/ctest/omega_ctest.py:
```diff
@@ -128,9 +128,9 @@ def download_meshes(config):
-        'cosine_bell_icos480_initial_state.230220.nc',
+        'ocean.QU.240km.omega_vars.251219.nc',
+        'PlanarPeriodic48x48.omega_vars.251219.nc',
+        'cosine_bell_icos480.omega_vars.250219.nc',
```
Need to be updated again.

xylar committed:
Rename outputs in forward runs

xylar committed:
Update conversion between pseudo- and geometric thickness

We no longer assume that `LayerThickness` exists (and that it
is pseudo-thickness).  Instead, we explicitly compute only
`PseudoThickness` and `SpecVol` when writing a model dataset
and compute `GeomLayerThickness` if needed when reading it back.

xylar committed:
Update conservation checks to use PseudoThickness

xylar committed:
Add function to get pseudothickness iter count

This is in anticipation of other code besides just the vertical
coordinate diagnostics needing this capability.

xylar committed:
Rework the z-tilde coordinate to be independent of z-star

Following the renaming of fields, trying to leverage the same
infrastructure is unnecessarily complicated and poteintially
limiting.

xylar committed:
Handle lack of `Time` in restingThickness --> refPseudoThickness

xylar committed:
Update CTest meshes

xylar committed:
Rename BottomDepth --> BottomGeomDepth

xylar committed:
Fix merry-go-round initial state conversion

xylar committed:
Add constant EOS to merry-go-round

xylar committed:
Rename to VerticalPseudoVelocity

xylar commented:
For the baseline comparison, I used the following script to rename baseline variables:
```python
#!/usr/bin/env python3
"""
Rename Omega variable names in baseline NetCDF files to match the
support-omega-pseudo-thickness branch renaming.

Run with --dry-run first to preview changes, then without to apply.
"""
import argparse
import sys
from pathlib import Path

import netCDF4

RENAMES = {
    'LayerThickness': 'PseudoThickness',
    'VerticalVelocity': 'VerticalPseudoVelocity',
    'TotalVerticalVelocity': 'TotalVerticalPseudoVelocity',
    'BottomDepth': 'BottomGeomDepth',
    'ZMid': 'GeomZMid',
    'ZInterface': 'GeomZInterface',
}


def rename_file(path, dry_run):
    try:
        with netCDF4.Dataset(path, 'r') as ds:
            to_rename = {
                k: v for k, v in RENAMES.items()
                if k in ds.variables and v not in ds.variables
            }
    except Exception as e:
        print(f'SKIP (cannot read): {path}: {e}', file=sys.stderr)
        return

    if not to_rename:
        return

    for old, new in to_rename.items():
        tag = '[dry-run] ' if dry_run else ''
        print(f'{tag}{path}: {old} -> {new}')

    if dry_run:
        return

    with netCDF4.Dataset(path, 'r+') as ds:
        for old, new in to_rename.items():
            ds.renameVariable(old, new)


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('baseline_dir', help='Baseline work directory')
    parser.add_argument(
        '--dry-run', action='store_true',
        help='Print what would be renamed without modifying files',
    )
    args = parser.parse_args()

    baseline = Path(args.baseline_dir)
    if not baseline.is_dir():
        sys.exit(f'Error: {baseline} is not a directory')

    for nc_file in sorted(baseline.rglob('*.nc')):
        rename_file(nc_file, args.dry_run)


if __name__ == '__main__':
    main()
```

xylar commented:
## Testing

### Polaris `omega_pr` suite

Note: the baseline is after running the script above to rename the output variables.

- Baseline workdir: `/lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260515/omega-pr-develop-var-rename`
- Baseline build: `/lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260515/omega-pr-develop/build`
- PR build: `/lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260517/omega-pr-pseudo-thickness/build`
- PR workdir: `/lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260517/omega-pr-pseudo-thickness`
- Machine: `chrysalis`
- Partition: `compute`
- Compiler: `intel`
- Build type: `Release`
- Log: `/lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260517/omega-pr-pseudo-thickness/polaris_omega_pr.o1218940`
- Result:
  - Diffs (1 of 9):
    - `ocean/column/horiz_press_grad/salinity_gradient`

The diffs in the `horiz_press_grad/salinity_gradient` test are tiny:
```
NormalVelocityTend   Time index: 0
0:  l1: 1.60245008015590e-12  l2: 5.50724425496013e-14  linf: 2.72861160749427e-15
  FAIL /lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260517/omega-pr-pseudo-thickness/ocean/column/horiz_press_grad/salinity_gradient/forward_500m/output.nc
```
I think it's just related to the z-tilde refactor.

### CTest unit tests:
- Machine: `chrysalis`
- Compiler: `intel`
- Build type: `Release`
- Failures (7 of 40):
  - `AUXSTATE_TEST`
  - `IOSTREAM_TEST`
  - `TENDENCIES_TEST`
  - `PGRAD_TEST`
  - `STATE_TEST`
  - `DRIVER_TEST`
  - `VERTCOORD_TEST`
- Log: `/lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260517/omega-pr-pseudo-thickness/build/ctests.log`

So I have some work to do here still...

xylar committed:
Update Omega CTest files

xylar commented:
## Updated testing

### CTest unit tests:

After updating the CTest files with the latest mapping (including `BottomGeomDepth`), things look great!

- Machine: `chrysalis`
- Compiler: `intel`
- Build type: `Release`
- Result: All tests passed
- Log: `/lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260518/omega-pr-pseudo-thickness/build/ctests.log`

xylar commented:
@brian-oneill, @cbegeman and @sbrus89, I'm hoping you can review this along with https://github.com/E3SM-Project/Omega/pull/327 (which @brian-oneill has already reviewed, thank you!).

@cbegeman, I'm hoping in particular to get your take on the validation mapping approach, where I override the `layerThickness` --> `GeomLayerThickness` mapping so validation instead happens on `PseudoThickness` for Omega tests.  If that bugs you, I'm fine with going back to validating `GeomLayerThickness`, it just seemed strange to validate a variable that requires significant alteration on the Polaris side.

cbegeman commented:
@xylar Does this PR fix https://github.com/E3SM-Project/polaris/issues/584? At least whether the merry-go-round case raises NaNs?

xylar commented:
@cbegeman, I haven't looked into it so I'm not sure either way.  I can take a look (probably when Chrysalis is back).

cbegeman commented on polaris/ocean/vertical/ztilde.py:
```diff
@@ -347,3 +348,283 @@ def geom_height_from_pseudo_height(
+    ).transpose('nCells', 'nVertLevels')
+    cell_mask = np.logical_and(above_bot_mask, valid)
+
+    # nonzero top index isn't supporeted at least for now
```
```suggestion
    # nonzero top index isn't supported at least for now
```

cbegeman commented on polaris/ocean/vertical/ztilde.py:
```diff
@@ -347,3 +348,283 @@ def geom_height_from_pseudo_height(
+        min_layer_thickness,
+    )
+
+    pseudo_column_thickness, max_level_cell = _alter_pseudo_column_thickness(
+        config, pseudo_column_thickness, ref_pseudo_depth_bot, max_level_cell
+    )
```
Am I correct in inferring that by this stage, the bottom pressure from `pseudo_column_thickness` and `RhoSw` is not necessarily equal to the Dataset's BottomPressure?

xylar replied:
That's a good point.  We may need to provide an updated `BottomPressure`, much as we would provide an updated `bottomDepth` for the z-star and z-level coordinates.

xylar replied:
Thinking about this a bit more, I think the problem it more complicated.  We probably do want to update `BottomPressure` here when we alter `pseudo_column_thickness` but we also have the complication that somehow we want to alter `BottomGeomDepth` as well or we will effectively end up increasing SSH instead of deepening the bathymetry as intended.  I will have to think about this tomorrow.

xylar replied:
I think this is fixed in 0afdcc2, but I need to test again.

cbegeman commented on polaris/ocean/conservation.py:
```diff
@@ -11,7 +11,7 @@ def compute_total_mass(ds_mesh, ds):
     ds = _reduce_dataset_time_dim(ds)
     area_cell = ds_mesh.areaCell
-    layer_thickness = ds.layerThickness
+    layer_thickness = _get_layer_thickness(ds)
```
Is this needed because we aren't using `open_model_dataset` for ds? I think we want to keep the call as-is and use `open_model_dataset` so that, for Omega, `PseudoThickness` is converted to `GeomLayerThickness`, which is then mapped to `layerThickness`

cbegeman replied:
For the mass conservation check in particular, we can do something like (untested pseudocode)
```suggestion
if `layerThickness` in ds.keys():
    if `density` in ds.keys():
        total_mass = 
            area_cell * (ds.density * ds.layerThickness).sum(dim='nVertLevels')
     else:  # consider attempting to compute density from T,S,p using EOS
         logger.warn('density not present in output, computing mass conservation assuming reference density')
         total_mass = rho_sw * (
             area_cell * ds.layerThickness.sum(dim='nVertLevels')
elif `PseudoThickness` in ds.keys():
     total_mass = rho_sw * (
        area_cell * ds.PseudoThickness.sum(dim='nVertLevels')
```

xylar replied:
I don't see why my change isn't correct.  For Omega, `RhoSw*PseudoThickness*AreaCell` is the mass, and for MPAS-Ocean `rho_sw*layerThickness*areaCell` is its version of the conserved mass, referenced to volume.  I think we don't want to use `GeomLayerThickness` when it simply isn't the right variable for Omega mass conservation.

xylar replied:
Maybe `_get_layer_thickness()` should be model based, rather than based on what's in the dataset, though.

xylar replied:
`ds.density * ds.layerThickness` is a messy way to recompute `PseudoThickness` when it should already be present.

cbegeman replied:
Yes, I was trying to do something that is model-specific. For MPAS-O, we shouldn't assume that the density = RhoSw. For Omega, we should just use PseudoThickness

xylar replied:
I think MPAS-O should use a constant density, since it really conserves volume, not mass.  In https://github.com/E3SM-Project/polaris/pull/494, we agreed that this would be the PCD value.

I'm happy to update `_get_layer_thickness(ds)` but be model-specific rather than variable-specific.  The main problem up to now has been that `config` is not available but I can fix that.

cbegeman replied:
Good point. I had forgotten that we're really checking volume conservation for MPAS-O, not mass. Let's just keep what you have.

cbegeman commented:
> @brian-oneill, @cbegeman and @sbrus89, I'm hoping you can review this along with [E3SM-Project/Omega#327](https://github.com/E3SM-Project/Omega/pull/327) (which @brian-oneill has already reviewed, thank you!).
> 
> @cbegeman, I'm hoping in particular to get your take on the validation mapping approach, where I override the `layerThickness` --> `GeomLayerThickness` mapping so validation instead happens on `PseudoThickness` for Omega tests. If that bugs you, I'm fine with going back to validating `GeomLayerThickness`, it just seemed strange to validate a variable that requires significant alteration on the Polaris side.

My big-picture question is whether we should be using `ZMid`, `ZInterface` as the variables to validate since both MPAS-O and Omega compute them every time step. The way it's written right now, we would have many calls to `geom_thickness_from_ds` for Omega during a test suite.

xylar commented:
> My big-picture question is whether we should be using `ZMid`, `ZInterface` as the variables to validate since both MPAS-O and Omega compute them every time step. The way it's written right now, we would have many calls to `geom_thickness_from_ds` for Omega during a test suite.

I feel like we should be validating the main prognostic variables whenever possible.  We could validate the geometric height but that feels to me like a diagnostic to look at in addition to the main prognostic variable, not a substitute for `layerThickness`/`PseudoThickness`.

cbegeman commented:
I think I might be not grasping what the validate variables override does. Are any changes needed so that `validate_vars` compares `layerThickness` for MPAS-O vs MPAS-O, and `PseudoThickness` for Omega vs. Omega, and `layerThickness`/`GeomLayerThickness` for Omega vs. MPAS-O?

xylar commented:
Without my changes, `validate_vars` compares `layerThickness` for MPAS-Ocean and `GeomLayerThickness` for Omega.  The latter seems like a poor choice to me and I think we should instead validate `PseudoThickness`.  That's the purpose of the override.

xylar commented:
> and layerThickness/GeomLayerThickness for Omega vs. MPAS-O?

I have never tried to use `validate_vars` to compare an MPAS-Ocean run with an Omega run if that's what you mean.  I don't think that would be particularly useful but maybe I haven't thought it through.  The override would definitely make such a comparison harder.

cbegeman commented:
I totally agree we should be comparing `PseudoThickness` for Omega vs. Omega. We can table the Omega vs. MPAS-O case. I think the override approach still makes sense. If we added things like converting SA to PS to `open_model_dataset` we would want to override that for validation too.

cbegeman commented on polaris/ocean/model/mpaso_to_omega.yaml:
```diff
@@ -37,17 +36,31 @@ variables:
 
   # state
-  layerThickness: LayerThickness
+  layerThickness: GeomLayerThickness
```
Is there a plan to make this an Omega variable (or is it already and I am missing it)? If not, I'm not really sure whether we'd change anything but I kind of conceptualized this file as mapping only between variables that actually exist in each model.

xylar replied:
Hmm, you are right.  Omega doesn't define this variable.

But I have some difficulty leaving `layerThickness` unconverted in Omega input files, which would be the consequence of removing this.  We would have `layerThickness` and `PseudoThickness` in Omega input files with potential confusion about what `layerThickness` means.  I guess we could convert `layerThickness` to `PseudoThickness` and then remove it?

xylar replied:
I guess I also feel like Omega probably *should* define `GeomLayerThickness`, even if just as a diagnostic variable.

I don't know if I feel like this PR or https://github.com/E3SM-Project/Omega/pull/327 are quite the right places to resolve this issue.  I'd prefer to make an issue about it and resolve it later.

cbegeman replied:
I'm fine with making an issue about it and resolving it later. Do you want to do that or do you want me to do that?

xylar replied:
I created https://github.com/E3SM-Project/Omega/issues/412

xylar replied:
It does seem like we could remove `GeomLayerThickness` here and map to `layerThickness` instead in `open_model_dataset()`.  This would make geometric layer thickness a variable constructed on the Polaris side rather than something Omega could ostensibly provide.  I think it's worth making the change.

xylar replied:
I went ahead with this approach in 9b4c67c.

cbegeman replied:
Looks great. Thanks!

cbegeman commented:
@xylar Did you consider changing `validate_vars` to something like `validate_state` where the set of variables to validate is fixed and set to the model's prognostic variables? Maybe this would be an option in addition to `validate_vars` which would be a custom list. Just wondering whether there's an option that is more intuitive than the override approach.

xylar commented:
> @xylar Did you consider changing `validate_vars` to something like `validate_state` where the set of variables to validate is fixed and set to the model's prognostic variables? Maybe this would be an option in addition to `validate_vars` which would be a custom list. Just wondering whether there's an option that is more intuitive than the override approach.

That hadn't occurred to me.  I agree it would be more intuitive. I asked Codex to look into it.  Here's what it found in terms of how `validate_vars` is currently used for the 4 prognostic variables `layerThickness`, `normalVelocity`, `temperature` and `salinity`, and whether a switch to `validate_state` would be safe:

| Category | Steps | Feasibility |
|---|---|---|
| Init steps omitting `normalVelocity` | `baroclinic_channel/init`, `barotropic_channel/init`, `ice_shelf_2d/init`, `internal_wave/init` | Mostly harmless. They write `normalVelocity`; most are zero. `barotropic_channel/init` actually sets nonzero velocity from config, so adding it is useful. |
| Init step validating only thickness | `barotropic_gyre/init` | Harmless: it writes constant `temperature`, constant `salinity`, and zero `normalVelocity`. |
| Forward steps omitting `salinity` only | `barotropic_channel/forward`, `overflow/forward`, `seamount/forward` | Likely harmless, especially for `overflow`/`seamount`, which output tracers. Verify Omega salinity output for `barotropic_channel` before switching. |
| Forward steps validating only thickness/velocity | `internal_wave/forward`, `external_gravity_wave/forward`, `single_column/ideal_age` | Probably works where output streams include tracers, but it broadens validation to fields that are often passive/constant. |
| Transport-style tests | `cosine_bell`, `sphere_transport`, `merry_go_round` | These are the most semantically intentional omissions. They validate velocity plus debug tracers. Full state probably works, but it may dilute what those tests are trying to check. |
| Not safe as-is | `barotropic_gyre/forward`, `inertial_gravity_wave/forward`, `manufactured_solution/forward` | These output streams omit tracers, so adding `temperature`/`salinity` would fail unless streams change. |

It looks like we could do this but it would require turning on some state outputs for various tests where they are currently omitted on purpose.

I have already been surprised about the amount of scope-creep that this PR has initiated.  I had hoped it would mostly be a simple name change but it has required refactoring z-tilde and fixing merry-go-round in addition to the validation override.

Maybe the best solution is to table this and just continue to validate `layerThickness` in this PR.

xylar commented:
@cbegeman, shall I remove the `validate_vars` override and (after this PR) make an issue that we want a way to validate `PseudoThickness` rather than `GeomLayerThickness` for Omega steps?

xylar committed:
Drop `GeomLayerThickness` Omega variable

We use `layerThickness` in Polaris as before, and drop
the mapping ot `GeomLayerThickness` as Omega doesn't define any
such variable.

xylar commented:
I've gone ahead and removed to the confusing `validate_vars` override.  We'll revisit this later.

cbegeman commented:
> I've gone ahead and removed to the confusing `validate_vars` override. We'll revisit this later.

Great. Thanks for scoping out this change. We can totally make it in a separate PR.

xylar committed:
Fix partial-cell handling in z-tilde vertical coordinate initialization

- After `_alter_pseudo_column_thickness` adjusts the column, recompute
  `pseudo_bottom_depth` and update `ds.BottomPressure` so `ZTildeInterface`
  is anchored correctly at the pseudo sea surface and downstream callers
  see a consistent effective bottom pressure.
- In the outer bottom-pressure iteration, scale from the post-adjustment
  `BottomPressure` rather than the raw input, and detect full-cell cycling
  (adjusted pressure unchanged between iterations) to break early with a
  warning instead of diverging.
- Set `bottomDepth` from the actual converged geometric water column
  thickness rather than the target bathymetry depth, ensuring SSH = 0
  at initialization when partial cells push the effective bottom shallower
  than the target.

xylar committed:
Fix typo

Co-authored-by: Carolyn Begeman <cbegeman@lanl.gov>

xylar commented:
My latest testing after switching back to validating `layerThickness` is that the conversion isn't happening correctly when `Step.validate_baselines()` is called.   A fix is still needed, and I think the cleanest one is to make `config` a require argument of `open_model_dataset()` and `write_model_dataset()`.  This is going to touch a lot of files again, but it's the only solution I can see that is philosophically correct for the Polaris framework: The framework doesn't know when a `config` is needed and when not, so it needs to pass to pass the `config` every time.

xylar committed:
Require `config` in model datasets I/O in framework

xylar committed:
Require `config` in model datasets I/O in ocean framework

xylar committed:
Require `config` in model datasets I/O in ocean tasks

xylar committed:
Require `config` in model datasets I/O in sea-ice tasks

xylar commented:
## Latest Testing

As always, in conjunction with https://github.com/E3SM-Project/Omega/pull/327

### CTest unit tests:
- Machine: `chrysalis`
- Compiler: `intel`
- Build type: `Release`
- Result: All tests passed
- Log: `/lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260520/omega-pr-pseudo-thickness/build/ctests.log`

### Polaris `omega_pr` suite
- Baseline workdir: `/lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260520/omega-pr-baseline`
- Baseline build: `/lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260520/omega-pr-baseline/build`
- PR build: `/lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260520/omega-pr-pseudo-thickness/build`
- PR workdir: `/lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260520/omega-pr-pseudo-thickness2`
- Machine: `chrysalis`
- Partition: `compute`
- Compiler: `intel`
- Build type: `Release`
- Log: `/lcrc/group/e3sm/ac.xylar/polaris_1.0/chrysalis/test_20260520/omega-pr-pseudo-thickness2/polaris_omega_pr.o1220199`
- Result:
  - Diffs (1 of 9):
    - `ocean/column/horiz_press_grad/salinity_gradient`

The baseline differences are the same as in earlier testing (https://github.com/E3SM-Project/Omega/pull/327#issuecomment-4476323516).

As before, the baseline is after running the script in https://github.com/E3SM-Project/polaris/pull/440#issuecomment-4470526878 to rename the output variables.

xylar commented:
@cbegeman, I would appreciate your input on the last 4 commits.  As I argued above (https://github.com/E3SM-Project/polaris/pull/440#issuecomment-4496847855), I think we need to bite the bullet and pass `config` every time we open or write a model dataset because the framework can't know when it's needed and when not.

cbegeman commented:
> @cbegeman, I would appreciate your input on the last 4 commits. As I argued above ([#440 (comment)](https://github.com/E3SM-Project/polaris/pull/440#issuecomment-4496847855)), I think we need to bite the bullet and pass `config` every time we open or write a model dataset because the framework can't know when it's needed and when not.

@xylar Thanks for making that change. I considered doing that when I introduced the argument and probably should have done it then.

xylar commented:
Great, thanks @cbegeman.

xylar commented:
> @xylar Does this PR fix https://github.com/E3SM-Project/polaris/issues/584? At least whether the merry-go-round case raises NaNs?

@cbegeman, unfortunately, no, this PR doesn't fix the NaN convergence rate in the merry-go-round tests (#584).  That will still need investigation.

cbegeman reviewed (approved):
Approving on the basis of visual inspection and @xylar 's latest testing. Thank you for your thoroughness here!

xylar committed:
Update Omega submodule to 54d85d8c4f

This bring in:
* https://github.com/E3SM-Project/Omega/pull/335
* https://github.com/E3SM-Project/Omega/pull/381
* https://github.com/E3SM-Project/Omega/pull/400
* https://github.com/E3SM-Project/Omega/pull/414
* https://github.com/E3SM-Project/Omega/pull/327

xylar commented:
@brian-oneill and @sbrus89, I'd like to merge this tomorrow.  Could you review however you think is best?  Perhaps a test on a machine other than Chrysalis and/or a compiler other than Chrysalis' intel?

xylar commented:
@cbegeman, thanks so much for giving this a thorough review and for the important decisions along the way!

mark-petersen commented:
I tested this PR with the current head of omega `develop` on perlmutter cpu and gpu using
```
/deploy.py --recreate
source load_polaris_pm-cpu_gnu_mpich.sh
polaris suite -c ocean -t omega_pr --model omega -p /pscratch/sd/m/mpeterse/runs/test_omega_cpu/build -w $n/omega-pr

/deploy.py --machine pm-gpu --recreate 
source load_polaris_pm-gpu_gnugpu_mpich.sh
polaris suite -c ocean -t omega_pr --model omega -p /pscratch/sd/m/mpeterse/runs/test_omega_gpu/build -w $n/omega-pr -m pm-gpu
```
where I already compiled omega and ran cTest in the `/pscratch/sd/m/mpeterse/runs/test_omega_?pu/build` directories. Those passed all tests!

I then tested cpu on Frontier in a similar way, 
```
./deploy.py --recreate
source load_polaris_frontier_craygnu_mpich.sh
polaris suite -c ocean -t omega_pr --model omega -p /lustre/orion/cli115/scratch/mpetersen/runs/test_omega_cpu/build -w $n/omega-pr
```

and had these tests fail:
```
Task Runtimes:
0:00:00 PASS ocean/column/horiz_press_grad/salinity_gradient
0:00:00 PASS ocean/planar/barotropic_gyre/munk/free-slip
0:00:00 PASS ocean/planar/manufactured_solution/convergence_both/default
0:00:00 PASS ocean/planar/manufactured_solution/convergence_both/del2
0:00:02 FAIL ocean/planar/manufactured_solution/convergence_both/del4
0:00:01 FAIL ocean/planar/merry_go_round/default
0:00:17 PASS ocean/planar/overflow/default
0:00:24 PASS ocean/spherical/icos/cosine_bell/decomp
0:00:03 PASS ocean/spherical/icos/cosine_bell/restart
```
It looks like the issue is `bottomDepth` versus `BottomGeomDepth`:
```
pwd
/lustre/orion/cli115/scratch/mpetersen/runs/n/omega-pr/ocean/planar/manufactured_solution/convergence_both/del4/forward/200km_300s

cat omega.log
[info] Omega using calendar type: No Leap
[info] VertCoord: Error while reading InitialVertCoord stream
[critical] [VertCoord.cpp:590] VertCoord: Error reading BottomGeomDepth from InitialVertCoord
[critical] [Error.cpp:43] Omega aborting

ncdump init.nc |grep ottom
	double bottomDepth(NCells) ;
```
but I see the other tests that Passed have `bottomDepth` in the `init.nc` as well. I don't know what the cause is. It is strange that it passes on perlmutter and not frontier.

brian-oneill commented:
Strange, I just tried the omega_pr suite on Frontier with a craygnu build of the `develop` branch:
```bash
./deploy.py --machine frontier --compiler craygnu
source ./load_polaris_frontier_craygnu_mpich.sh
polaris suite -c ocean -t omega_pr --model omega -p  /lustre/orion/proj-shared/cli115/boneill/dev/gnuCPU -w /lustre/orion/proj-shared/cli115/boneill/polaris/omega_craygnu/
```
and all the tests all passed:
```bash
Task Runtimes:
0:00:57 PASS ocean/column/horiz_press_grad/salinity_gradient
0:00:07 PASS ocean/planar/barotropic_gyre/munk/free-slip
0:01:07 PASS ocean/planar/manufactured_solution/convergence_both/default
0:00:53 PASS ocean/planar/manufactured_solution/convergence_both/del2
0:00:52 PASS ocean/planar/manufactured_solution/convergence_both/del4
0:00:13 PASS ocean/planar/merry_go_round/default
0:00:25 PASS ocean/planar/overflow/default
0:00:24 PASS ocean/spherical/icos/cosine_bell/decomp
0:00:05 PASS ocean/spherical/icos/cosine_bell/restart
Total runtime: 0:05:03
PASS: All passed successfully!
```

The `bottomDepth` to `BottomGeomDepth` rename shows up in that `init.nc` for me
```bash

cd ocean/planar/manufactured_solution/convergence_both/del4/forward/200km_300s
ncdump init.nc |grep ottom
	double BottomGeomDepth(NCells) ;
```

Ran the suite on Frontier with craygnu-mphipcc as well and all tests passed.

brian-oneill commented on polaris/ocean/model/mpaso_to_omega.yaml:
```diff
@@ -37,17 +36,23 @@ variables:
   normalVelocity: NormalVelocity
-
-  vertVelocityTop: VerticalVelocity
+  vertVelocityTop: VerticalPseudoVelocity
```
Should there be a similar note for `VerticalPseudoVelocity` here, since `vertVelocityTop` is not equivalent in the same way `layerThickness` and `PseudoThickness` are not equivalent?

xylar replied:
Excellent point!

xylar replied:
Done in caeb1fd.

brian-oneill reviewed (approved):
LGTM, I don't have a deep grasp on the inner workings of Polaris, but from what I can tell all the instances of `LayerThickness` have been updated to `PseudoThickness`, just one comment. Based on successfully running the omega_pr suite on Frontier, I approve.

xylar commented on polaris/ocean/model/mpaso_to_omega.yaml:
```diff
@@ -37,17 +36,23 @@ variables:
   normalVelocity: NormalVelocity
-
-  vertVelocityTop: VerticalVelocity
+  vertVelocityTop: VerticalPseudoVelocity
```
```suggestion
  # Note: vertVelocityTop in MPAS-Ocean has no Omega equivalent and
  #       VerticalPseudoVelocity in Omega has no MPAS-Ocean equivalent
```

xylar committed:
Remove VerticalPseudoVelocity from map

There is no MPAS-Ocean equivalent.

xylar commented:
@mark-petersen, could you try again?  It seems like maybe the Polaris branch you were testing with wasn't this one, given that things work for Brian.

xylar commented:
@mark-petersen and @sbrus89, I think 2 approvals is sufficient.  I'm going to go ahead here.  @mark-petersen, your successful tests on Perlmutter are also a good indication that things look as they should.  @brian-oneill's successful test on Frontier really suggests to me that something was misconfigured somehow in @mark-petersen's test but we can revisit again if it turns out to be reproducable.

mark-petersen commented:
Yes, I reran in a clean work directory and all tests passed on Frontier. The problem is that I reused a previous work directory from before this PR, and it did not re-run the init step for those, so it had the old names in it. Sorry for the confusion.

xylar commented:
Glad we go that cleared up!
